smart_open/ssh.py

```diff
@@ -106,4 +106,8 @@ def open(path, mode='r', host=None, user=None, password=None, port=DEFAULT_PORT,
 
     conn = _connect(host, user, port, password, transport_params)
     sftp_client = conn.get_transport().open_sftp_client()
-    return sftp_client.open(path, mode)
+    file_obj = sftp_client.open(path, mode)
+    is_pipelining_enabled = transport_params.get('write_pipelining', True)
```

Comment:
You shouldn't rely on `transport_params` here. That's a poorly named parameter here, it should really be called `connect_params` or something, because it's only for the connect function.

You should add a new keyword parameter `write_pipelining` and document it in the docstring.

Reply:
I don't get it. ssh.connect keyword args are passed through 'connect_kwargs' key in `transport_params` and this is IMHO fine - this allows for introducing other parameters directly in transport_params without affecting ssh.connect arguments (like this `write_pipelining` parameter). I can add `write_pipelining` keyword parameter to ssh.open function signature and initialize it while calling ssh.open (taking value from transport_params dict), but I don't see any advantage of that (this will introduce more backend-related code to smart_open_lib). Could you describe more precisely how do you see it?

Reply:
The advantage is discovery of your new feature. If you add a new parameter, people will find out about it via the doc string. Otherwise, no-one will know about it.

The existing handling of transport Params by the ssh submodule is broken (look at the current docstring). We will have to fix that in a separate PR.

Reply:
> The advantage is discovery of your new feature. If you add a new parameter, people will find out about it via the doc string. Otherwise, no-one will know about it.
> 
So maybe backend should be called like `ssh.open(..., **transport_params)`? This way we can have separate keyword parameters in backend's open function (with docstring documentation) and don't have this backend-specific parameter handling on upper level.

> The existing handling of transport Params by the ssh submodule is broken (look at the current docstring). We will have to fix that in a separate PR.

I thought it is implemented as you described here: https://github.com/RaRe-Technologies/smart_open/pull/292#issuecomment-486478592

Reply:
> So maybe backend should be called like `ssh.open(..., **transport_params)`?

Yes, this.